extendr-api/src/wrapper/array.rs

```diff
@@ -0,0 +1,339 @@
+//! Wrappers for matrices with deferred arithmetic.
+
+use super::scalar::Scalar;
+use super::*;
+use std::{
+    marker::PhantomData,
+    ops::{Index, IndexMut},
+};
+
+/// Wrapper for creating and using matrices and arrays.
+///
+/// ```
+/// use extendr_api::prelude::*;
+/// test! {
+///     let matrix = RArray::from_slice([1,2,3,4,5,6], [3, 2]);
+///     let robj = r!(matrix);
+///     assert_eq!(robj.is_matrix(), true);
+///     assert_eq!(robj.nrows(), 3);
+///     assert_eq!(robj.ncols(), 2);
+/// }
+/// ```
+#[derive(Debug, PartialEq)]
+pub struct RArray<T, D> {
+    /// Owning Robj (probably should be a Pin).
+    robj: Robj,
+
+    /// Dimensions of the array.
+    dim: D,
+
+    phantom: PhantomData<T>,
+}
+
+pub type RMatrix<T> = RArray<T, [usize; 2]>;
+
+impl<'a, T, D> RArray<T, D>
+where
+    Robj: AsTypedSlice<'a, T>,
+    Self: 'a,
+{
+    // Make a new RArray type.
+    // This function doesn't insert a dim attribute to the Robj.
+    pub(crate) fn new(robj: Robj, dim: D) -> Self {
+        Self {
+            robj,
+            dim,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Get the underlying data from this RArray.
+    pub fn data(&'a self) -> &'a [T] {
+        self.robj.as_typed_slice().unwrap()
+    }
+
+    /// Get the dimensions for this array.
+    pub fn dim(&self) -> &D {
+        &self.dim
+    }
+
+    /// Make a new RArray type.
+    /// This function inserts a dim attribute to the Robj.
+    pub fn from_slice<P>(slice: P, dims: D) -> Result<Self>
+    where
+        P: AsRef<[T]>,
+        T: ToVectorValue + Clone + 'a,
+        D: AsRef<[usize]> + Clone,
+    {
+        let dims_ref = dims.as_ref();
+        let slice_ref = slice.as_ref();
+        let robj = slice_ref.iter().cloned().collect_robj();
+
+        let prod = dims_ref.iter().product::<usize>();
+        if prod != robj.len() {
+            return Err(Error::Other(format!(
+                "The vector length ({}) does not match the length implied by the dimensions ({})",

```

Comment:
I saw this error twice in the codebase, I believe it deserves its own error type with perhaps two parameters and this string moved to `impl Debug`

Reply:
I was trying to implement a generic `ValuesMismatch` error where it accepts 2 values that implement the debug trait but the partialeq derive for Error blocked it... I think it would be better to create an issue in order to implement it another time...
I implemented a `DimensionMismatch` error anyway, but I ended up not too happy about it since it is too specialized...